cpp/src/arrow/python/decimal.cc

```diff
@@ -184,14 +184,15 @@ Status DecimalMetadata::Update(int32_t suggested_precision, int32_t suggested_sc
 }
 
 Status DecimalMetadata::Update(PyObject* object) {
-  DCHECK(PyDecimal_Check(object)) << "Object is not a Python Decimal";
+  bool is_decimal = PyDecimal_Check(object);
```

Comment:
I don't think it's ok to do this in optimized build. `DecimalMetadata` expects you to pass a decimal object. @cpcloud may confirm.

Reply:
This isn't necessary because I added a check before calling `Update` but it does prevent a seg fault if for some reason it's called with non Decimal objects - which is not nice to get.  If it's hurts an optimization though, I can remove it.

Reply:
Right now we are doing the check twice in optimized builds, which is not nice IMHO. `DecimalMetadata::Update` is a private API so it's up to the caller to provide appropriate input.

Reply:
So you mean remove `PyDecimal_Check` all together?  This is only called when the type is not specified by the user and then yes, it will end doing 2 passes over the objects and checks both times if they are decimal.  It might be possible to do less checks on the second pass if we keep a list of which ones are decimal objects, but I'm not sure that would be worth it.

Reply:
Fair enough, we can optimize later if we find it too slow. The conversion itself is very slow anyway :-)